File: backend/internal/application/shuffle.go
package application

import "math/rand/v2"

// FisherYatesShuffle returns a new slice containing a uniformly random
// permutation of items. The input slice is not mutated.
func FisherYatesShuffle[T any](items []T, rng *rand.Rand) []T {
	result := make([]T, len(items))
	copy(result, items)
	for i := len(result) - 1; i > 0; i-- {
		j := rng.IntN(i + 1)
		result[i], result[j] = result[j], result[i]
	}
	return result
}

File: frontend/src/lib/shuffle.ts
/** A random source returning a float in [0, 1). Injectable for deterministic tests. */
export type RandomFn = () => number;

/**
 * Fisher–Yates shuffle. Returns a new array; the input is not mutated.
 * Every permutation is equally likely given a uniform random source.
 */
export function fisherYatesShuffle<T>(
  items: readonly T[],
  random: RandomFn = Math.random,
): T[] {
  const result = [...items];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    const a = result[i] as T;
    const b = result[j] as T;
    result[i] = b;
    result[j] = a;
  }
  return result;
}
